cmd/kubelet/app/options/globalflags.go

```diff
@@ -0,0 +1,167 @@
+	_ "github.com/google/cadvisor/machine"
+	_ "github.com/google/cadvisor/manager"
+	_ "github.com/google/cadvisor/storage"
+	_ "k8s.io/kubernetes/pkg/credentialprovider/azure"
+	_ "k8s.io/kubernetes/pkg/credentialprovider/gcp"
+)
+
+// AddGlobalFlags explicitly registers flags that libraries (glog, verflag, etc.) register
+// against the global flagsets from "flag" and "github.com/spf13/pflag".
+// We do this in order to prevent unwanted flags from leaking into the Kubelet's flagset.
+func AddGlobalFlags(fs *pflag.FlagSet) {
+	addGlogFlags(fs)
+	addCadvisorFlags(fs)
+	verflag.AddFlags(fs)
+	logs.AddFlags(fs)
+}
+
+// normalize replaces underscores with hyphens
+// we should always use hyphens instead of underscores when registering kubelet flags
+func normalize(s string) string {
+	return strings.Replace(s, "_", "-", -1)
+}
+
+// register adds a flag to local that targets the Value associated with the Flag named globalName in global
+func register(global *flag.FlagSet, local *pflag.FlagSet, globalName string) {
+	if f := global.Lookup(globalName); f != nil {
+		f.Name = normalize(f.Name)
+		local.AddFlag(pflag.PFlagFromGoFlag(f))
+	} else {
+		panic(fmt.Sprintf("failed to find flag in global flagset (flag): %s", globalName))
+	}
+}
+
+// pflagRegister adds a flag to local that targets the Value associated with the Flag named globalName in global
+func pflagRegister(global, local *pflag.FlagSet, globalName string) {
+	if f := global.Lookup(globalName); f != nil {
+		f.Name = normalize(f.Name)
+		local.AddFlag(f)
+	} else {
+		panic(fmt.Sprintf("failed to find flag in global flagset (pflag): %s", globalName))
+	}
+}
+
+// registerDeprecated registers the flag with register, and then marks it deprecated
+func registerDeprecated(global *flag.FlagSet, local *pflag.FlagSet, globalName, deprecated string) {
+	register(global, local, globalName)
+	local.Lookup(normalize(globalName)).Deprecated = deprecated
+}
+
+// pflagRegisterDeprecated registers the flag with pflagRegister, and then marks it deprecated
+func pflagRegisterDeprecated(global, local *pflag.FlagSet, globalName, deprecated string) {
+	pflagRegister(global, local, globalName)
+	local.Lookup(normalize(globalName)).Deprecated = deprecated
+}
+
+// addCredentialProviderFlags adds flags from k8s.io/kubernetes/pkg/credentialprovider
+func addCredentialProviderFlags(fs *pflag.FlagSet) {
+	// lookup flags in global flag set and re-register the values with our flagset
+	global := pflag.CommandLine
+	local := pflag.NewFlagSet(os.Args[0], pflag.ExitOnError)
+
+	// Note this is deprecated in the library that provides it, so we just allow that deprecation
+	// notice to pass through our registration here.
+	pflagRegister(global, local, "google-json-key")
+	// TODO(mtaufen): This is not a static file, so it's not quite as straightforward as --google-json-key.
+	// We need to figure out how ACR users can dynamically provide pull credentials before we can deprecate this.
```

Comment:
can you spawn an issue for this, tagging @smarterclayton and the azure cloudprovider owners?

Reply:
#58034